Design/histogram_type.md

```diff
@@ -0,0 +1,108 @@
+- Hide differences between histogram and distribution data.
+  - Internal storage type could be either, does not need to be known at the interface level?
+  - Arithmetic operations should do the right thing for any data type automatically.
+- Hide error handling.
+  - For simple operations with histograms algorithm authors should not need to know how to propagate errors.
+- Enable unit testing of key concepts in Mantid.
+  - Arithmetic operations with histograms.
+  - Propagation of errors.
+
+
+
+## Implementation
+
+#### Basics
+
+The details of the implementation have not been worked out yet.
+The basic concept seems simple enough:
+
+- Add one class for a histogram with an interface that provides standard operations for histogram data.
+  - Note that we do not introduce a separate type for histogram and distribution data. This distinction would be only internal, and access to either histogram or distribution data is provided via public interface methods. The goal is to eliminate this distinction from as much of the code as possible.
+- Add a few classes for `X`, `Y`, and `E` (etc.) that would mainly be used internally in the histogram class and for direct access to the underlying data. If these inherit from `std::vetor<double>`, they could also be used to keep the legacy interface for `ISpectrum` alive until the roll-out is complete.
+
+#### Details
+
+- It is probably useful to introduce two types for vectors of standard deviation and variance, instead of the current double use of `E`.
+- Similarly we should have distinct internal types for histogram and distribution data, in case our implementation relies on this distinction. Jon pointed out that it may make most sense to default to distribution data, as OpenGenie did by default.
+- The current copy-on-write (COW) mechanism seems useful and should be preserved.
+  - The `typedef` `MantidVecPtr` for a `cow_ptr` is useless and confusing and should be removed.
+  - We should consider extending the COW mechanism and make sure sharing is preserved more than it is now. For example, what can we do to maintain sharing of `X` for all histograms in a workspace where `X` is modified in an identical way? The COW mechanism as it is now should break down in that case. Does it need to be handled at the workspace level, or can we deal with this here? Having parametric `X` could circumvent that in certain cases (see outlook).
+
+#### Interface
+
+- At least initially not all cases for operations with histograms will be covered by the histogram type or free functions for histograms. Therefore we probably need to provide a direct interface to the underlying data as well.
+- If we want to hide things like error handling, is there a decent way of making this extensible for new operations? Are operators/functions declared as `friend` in our histogram data type? Are they members? Can we allow the use of `std::transform` on, e.g., the `Y` data? How to transform errors?
+
```

Comment:
The other question is whether we need more direct access to the data for things like plotting/visualization. `Qwt` has abstractions for the data source so this can probably be mitigated in those classes but it's something we need to make sure we understand.